### PR TITLE
better precision of value_set on unions

### PR DESCRIPTION
**Pull request by kroening**
This is motivated by the linked list implementation in FreeRTOS.

### CODE REVIEW

**Comment by tautschnig** [comment]
What is the relationship between this one and #2854?

File: src/pointer-analysis/value_set.cpp
Comment (tautschnig) [inline src/pointer-analysis/value_set.cpp]: Doesn't the simplifier do this? If it doesn't, should it do so? (I note that we do call upon the simplifier anyway, so if the simplifier were taught about it then it should be addressed in here in all cases?)

**Comment by kroening** [comment]
Relationship to #2854: They both concern unions, but appear to be different.

File: src/pointer-analysis/value_set.cpp
Comment (kroening) [reply src/pointer-analysis/value_set.cpp]: Will look into it!

Comment (tautschnig) [reply src/pointer-analysis/value_set.cpp]: Done.

**Review by tautschnig — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 0c8ef661845c04abf5d71a98185f9c983c2648b2).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

File: src/util/simplify_expr_struct.cpp
Comment (smowton) [inline src/util/simplify_expr_struct.cpp]: Const ref?

File: regression/cbmc/union11/union_list.c
Comment (smowton) [inline regression/cbmc/union11/union_list.c]: Comment inline about whether the single-item union is important or not, since that's pretty unusual afaik

File: regression/cbmc/union11/union_list.c
Comment (smowton) [inline regression/cbmc/union11/union_list.c]: Unused in the test case? Is it important that neither struct member is at zero offset? If so comment inline about why

File: regression/cbmc/union11/union_list.c
Comment (smowton) [inline regression/cbmc/union11/union_list.c]: Why a single member struct instead of simply inlining this?

File: src/pointer-analysis/value_set.cpp
Comment (smowton) [inline src/pointer-analysis/value_set.cpp]: If I'm reading this right the ID_unknown case used to be silently ignored? I suspect this increase in pessimism / improvement in soundness will have side effects, consider doing that as a separate PR since that seems orthogonal to simplifying unions?

File: src/util/simplify_expr_struct.cpp
Comment (tautschnig) [reply src/util/simplify_expr_struct.cpp]: Only once #2987 has been merged.

**Comment by tautschnig** [comment]
I would like to wait for #2987 to be merged before doing further work here.

File: regression/cbmc/union11/union_list.c
Comment (tautschnig) [reply regression/cbmc/union11/union_list.c]: It's crucial to make this pass/fail. I have added both an inline comment and information to the commit message.

File: regression/cbmc/union11/union_list.c
Comment (tautschnig) [reply regression/cbmc/union11/union_list.c]: We could have more members, just don't need more to demonstrate the problem/the fix.

**Comment by tautschnig** [comment]
@smowton All comments addressed.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 61ffc48a522fa725d90ce7519d5039c372c57d76).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85635549
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: f2f2c601c4b08110aac3c773199580d13fbc3a2b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/85636056

**Review by smowton — APPROVED** [review]

**Comment by kroening** [comment]
@tautschnig Indeed much better to have this in the simplifier.

**Pull request merged**
